### PR TITLE
Upgrade Hibernate ORM to 6.2.5.Final and Hibernate Reactive to 2.0.1.Final 

### LINKED ISSUE #33779: Quarkus 3/Hibernate 6 - Exception when access ElementCollection from an Entity with an embedded composite Key with associated Object
### Describe the bug

I have an Entity with an EmbeddedId and an ElementCollection: 
```
@Entity()
public class MyEntity {
    @EmbeddedId
    EmbeddableId id;

    public String name;

    @ElementCollection
    public Set<EmbeddableElement> elementCollection = new HashSet<>();
}
```
```
@Embeddable
public class EmbeddableElement {
    public String name;
}
```

The embedded id has an composite key
```
@Embeddable
public class EmbeddableId  {
    public String id1;
    @ManyToOne
    public PartOfIdEntity id2;
}
```
```
@Entity()
public class PartOfIdEntity  {
    @Id
    public Long id;
}
```

### Expected behavior

No Exception when accessing the ElementCollection like in Quarkus 2.*

### Actual behavior

I get an Exception

Now if i access the elementCollection of MyEntity i get following exception: 
```
java.lang.NullPointerException: Cannot invoke "org.hibernate.sql.results.graph.entity.EntityInitializer.getNavigablePath()" because "firstEntityInitializer" is null
        at org.hibernate.sql.results.graph.entity.internal.BatchEntityInsideEmbeddableSelectFetchInitializer.getRootEmbeddablePropertyName(BatchEntityInsideEmbeddableSelectFetchInitializer.java:190)
        at org.hibernate.sql.results.graph.entity.internal.BatchEntityInsideEmbeddableSelectFetchInitializer.<init>(BatchEntityInsideEmbeddableSelectFetchInitializer.java:57)
        at org.hibernate.sql.results.graph.entity.internal.EntitySelectFetchInitializerBuilder.createInitializer(EntitySelectFetchInitializerBuilder.java:58)
        at org.hibernate.sql.results.graph.entity.internal.EntityFetchSelectImpl.buildEntitySelectFetchInitializer(EntityFetchSelectImpl.java:92)
        at org.hibernate.sql.results.graph.entity.internal.EntityFetchSelectImpl.lambda$createAssembler$0(EntityFetchSelectImpl.java:70)
        at org.hibernate.sql.results.internal.ResultsHelper$1.resolveInitializer(ResultsHelper.java:108)
        at org.hibernate.sql.results.graph.entity.internal.EntityFetchSelectImpl.createAssembler(EntityFetchSelectImpl.java:67)
        at org.hibernate.sql.results.graph.embeddable.AbstractEmbeddableInitializer.initializeAssemblers(AbstractEmbeddableInitializer.java:121)
        at org.hibernate.sql.results.graph.embeddable.AbstractEmbeddableInitializer.<init>(AbstractEmbeddableInitializer.java:107)
        at org.hibernate.sql.results.graph.embeddable.internal.EmbeddableResultInitializer.<init>(EmbeddableResultInitializer.java:22)
        at org.hibernate.sql.results.graph.embeddable.internal.EmbeddableForeignKeyResultImpl.lambda$createResultAssembler$0(EmbeddableForeignKeyResultImpl.java:101)
        at org.hibernate.sql.results.internal.ResultsHelper$1.resolveInitializer(ResultsHelper.java:108)
        at org.hibernate.sql.results.graph.embeddable.internal.EmbeddableForeignKeyResultImpl.createResultAssembler(EmbeddableForeignKeyResultImpl.java:98)
        at org.hibernate.sql.results.graph.collection.internal.CollectionDomainResult.lambda$createResultAssembler$0(CollectionDomainResult.java:98)
        at org.hibernate.sql.results.internal.ResultsHelper$1.resolveInitializer(ResultsHelper.java:108)
        at org.hibernate.sql.results.graph.collection.internal.CollectionDomainResult.createResultAssembler(CollectionDomainResult.java:94)
        at org.hibernate.sql.results.jdbc.internal.StandardJdbcValuesMapping.resolveAssemblers(StandardJdbcValuesMapping.java:53)
        at org.hibernate.sql.results.internal.ResultsHelper.createRowReader(ResultsHelper.java:78)
        at org.hibernate.sql.results.internal.ResultsHelper.createRowReader(ResultsHelper.java:64)
        at org.hibernate.sql.exec.internal.JdbcSelectExecutorStandardImpl.doExecuteQuery(JdbcSelectExecutorStandardImpl.java:341)
        at org.hibernate.sql.exec.internal.JdbcSelectExecutorStandardImpl.executeQuery(JdbcSelectExecutorStandardImpl.java:168)
        at org.hibernate.sql.exec.internal.JdbcSelectExecutorStandardImpl.list(JdbcSelectExecutorStandardImpl.java:93)
        at org.hibernate.sql.exec.spi.JdbcSelectExecutor.list(JdbcSelectExecutor.java:31)
        at org.hibernate.loader.ast.internal.SingleIdLoadPlan.load(SingleIdLoadPlan.java:144)
        at org.hibernate.loader.ast.internal.SingleIdLoadPlan.load(SingleIdLoadPlan.java:104)
        at org.hibernate.loader.ast.internal.CollectionLoaderBatchKey.load(CollectionLoaderBatchKey.java:116)
        at org.hibernate.persister.collection.AbstractCollectionPersister.initialize(AbstractCollectionPersister.java:668)
        at org.hibernate.event.internal.DefaultInitializeCollectionEventListener.onInitializeCollection(DefaultInitializeCollectionEventListener.java:75)
        at org.hibernate.event.service.internal.EventListenerGroupImpl.fireEventOnEachListener(EventListenerGroupImpl.java:127)
        at org.hibernate.internal.SessionImpl.initializeCollection(SessionImpl.java:1697)
        at org.hibernate.collection.spi.AbstractPersistentCollection.lambda$initialize$3(AbstractPersistentCollection.java:617)
        at org.hibernate.collection.spi.AbstractPersistentCollection.withTemporarySessionIfNeeded(AbstractPersistentCollection.java:265)
        at org.hibernate.collection.spi.AbstractPersistentCollection.initialize(AbstractPersistentCollection.java:615)
        at org.hibernate.collection.spi.AbstractPersistentCollection.read(AbstractPersistentCollection.java:136)
        at org.hibernate.collection.spi.AbstractPersistentCollection.lambda$readSize$0(AbstractPersistentCollection.java:163)
        at org.hibernate.collection.spi.AbstractPersistentCollection.withTemporarySessionIfNeeded(AbstractPersistentCollection.java:265)
        at org.hibernate.collection.spi.AbstractPersistentCollection.readSize(AbstractPersistentCollection.java:148)
        at org.hibernate.collection.spi.PersistentSet.size(PersistentSet.java:151)
        at example.MyEntityTest.test(MyEntityTest.java:47)

```

### How to Reproduce?

1. Persist some data 
```
PartOfIdEntity partOfIdEntity = new PartOfIdEntity();
            partOfIdEntity.id = 1L;
            entityManager.persist(partOfIdEntity);

            MyEntity myEntity = new MyEntity();
            myEntity.id = new EmbeddableId();
            myEntity.id.id1 = "123";
            myEntity.id.id2 = partOfIdEntity;

            EmbeddableElement el = new EmbeddableElement();
            el.name = "abc";
            myEntity.elementCollection.add(el);
            entityManager.persist(myEntity);
```
2. Close the transaction
3. Access the data:
```
 MyEntity entityFromSelect = em.createQuery("SELECT e FROM MyEntity e", MyEntity.class).getSingleResult();
        entityFromSelect.elementCollection.size();

        // Cannot invoke "org.hibernate.sql.results.graph.entity.EntityInitializer.getNavigablePath()" because "firstEntityInitializer" is null:
        assertEquals(1, entityFromSelect.elementCollection.size());
```

### Output of `uname -a` or `ver`

_No response_

### Output of `java -version`

openjdk version "17.0.7" 2023-04-18 OpenJDK Runtime Environment Temurin-17.0.7+7 (build 17.0.7+7) OpenJDK 64-Bit Server VM Temurin-17.0.7+7 (build 17.0.7+7, mixed mode, sharing)

### GraalVM version (if different from Java)

_No response_

### Quarkus version or git rev

3.1

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

_No response_

### Additional information

First i assumed it was a hibernate issue, but i can´t reproduce it with pure hibernate:
[hibernate-orm-6.zip](https://github.com/quarkusio/quarkus/files/11632131/hibernate-orm-6.zip)

Only with Quarkus 3.*
[quarkus-hibernate-issue.zip](https://github.com/quarkusio/quarkus/files/11632133/quarkus-hibernate-issue.zip)

### PR DESCRIPTION
**Pull request by DavideD**
* Hibernate ORM to 6.2.5.Final
* Hibernate Reactive to 2.0.1.Final

Sorry, I've deleted the previous PR (#34058) by mistake

### CODE REVIEW

**Review by Sanne — APPROVED** [review]

**Comment by famod** [comment]
Can this be backported to 3.1.3?

**Comment by gsmet** [comment]
@famod for now it cannot be merged due to failures. The upgrade will be backported if we end up being able to merge it.

File: extensions/smallrye-reactive-messaging-kafka/deployment/pom.xml [hunk 1]
```diff
@@ -42,6 +42,10 @@
             <groupId>io.quarkus</groupId>
             <artifactId>quarkus-jackson-deployment</artifactId>
         </dependency>
+        <dependency>
+            <groupId>io.quarkus</groupId>
+            <artifactId>quarkus-hibernate-orm-deployment</artifactId>
+        </dependency>
```
Comment (gsmet) [inline extensions/smallrye-reactive-messaging-kafka/deployment/pom.xml:48]: I don't think you can add a deployment dependency to an extension without also adding the runtime implementation. And that is going to be an issue.

Looks like this build item should be an SPI module (unfortunately).

Comment (yrodiere) [reply extensions/smallrye-reactive-messaging-kafka/deployment/pom.xml:48]: Ok, but there *are* dependencies to the runtime in the runtime module, they're just optional:

https://github.com/quarkusio/quarkus/blob/ba8c0fef65051e2743800a67f2b92cb5c59fc32c/extensions/smallrye-reactive-messaging-kafka/runtime/pom.xml#L34-L43

I'll move this to SPI, but I wonder if the requirements should be adjusted to take into account optional dependencies.

**Comment by yrodiere** [comment]
I confirmed the Spring Data failure as a regression in ORM (see https://hibernate.zulipchat.com/#narrow/stream/132094-hibernate-orm-dev/topic/Bytecode.20enhancement.20and.20embedded.20ID ; I'll open an issue later).

@gsmet Do you want me to find a workaround and merge anyway, or wait for the next version of ORM that fixes this regression?

**Comment by yrodiere** [comment]
I filed https://hibernate.atlassian.net/browse/HHH-16832 for the ORM bytecode enhancement bug. I will attempt to workaround the bug in our test suite for now.

**Comment by yrodiere** [comment]
Alright. I moved `AdditionalJpaModelBuildItem` to the SPI module to pass the automated check, and worked around https://hibernate.atlassian.net/browse/HHH-16832 in Spring Data JPA tests. The Hibernate ORM team is going to investigate the HHH-16832 in its next sprint.

**Comment by yrodiere** [comment]
Test failures are the usual suspects (flaky tests).

@gsmet still ok to merge the workaround?

**Comment by ChMThiel** [comment]
when will hibernate 6.2.5 will be available? We need this fix, because of this [hibernate-bug](https://hibernate.atlassian.net/browse/HHH-16612) we are not able to update.

**Comment by yrodiere** [comment]
> @gsmet still ok to merge the workaround?

**Pull request merged**